compiler/passes/insertWideReferences.cpp

```diff
@@ -18,11 +18,148 @@
  */
 
 //
-// Insert wide references
+// Insert Wide References (IWR)
 //
 // See notes at the top of parallel.cpp for background on heap
 // allocation in this pass
 //
+// The entry point for this pass is the function 'insertWideReferences'
+// located at the bottom of this file.
+//
+// This pass inserts the wide pointers required for communication across
+// locales.
+//
+// Note: Prior to ~July 2015 there were two passes that performed the same
+// duties of this single pass:
+//   - insertWideReferences: Simply widened everything by default
+//   - narrowWideReferences: Selectively narrowed variables
+//
+// --------------------------------------------------
+//
+// Some simple terminology used in this pass:
+//
+// - "wide" : Something that may be remote, type will differ from local vars.
+//
+// - "narrow" : A local variable.
+//
+// - "val" : Refers to the '_val' field in a reference type.
+//
+// --------------------------------------------------
+//
+// A remote variable is represented by a simple struct with two fields:
+//   - A chpl_localeID_t representing the locale where the data lives
+//   - A pointer to data on the locale
+//
+// Runtime calls are inserted by the compiler to read and write such variables.
+//
+// This pass is responsible for the creation of these types.
+// See buildWideClasses and buildWideRefMap.
+//
+// --------------------------------------------------
+//
+// We start by identifying variables that are assumed to be remote. These
+// variables are usually used in ways where it's currently hard to tell if they
+// can ever be local, so for correctness' sake we assume wideness.
+// See addKnownWides() for more.
+//
+// From this small set, we then propagate wideness throughout the AST. Here are
+// a couple of simple cases:
+//
+// - move A, B
```

Comment:
I don't think this is true in all cases in the current implementation.  However, for the direction in which we are heading, it is OK to assume this for now -- mainly because it is difficult to isolate the cases where it is not true.
Those cases are opportunities for optimization: In this revised implementation, wideness is still propagated where it need not necessarily be.
The reason for the current ambiguity is that the backend will insert a dereference as needed, so that a move primitive can be translated into valid C code.  On a parallel branch, I am working to ensure that we are not relying on this normalization at codeGen time.
Back to the main point: If the operands of the MOVE are both pointer (references), then wideness propagates from B to A as stated above.  When derefs are inserted early (as on my branch), then this will be the *only* case; then, the above statement will always be true.